Http monitoring for windows

ISauve opened the pull request:
### What does this PR do?

Adds http monitoring for windows.

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Describe how to test your changes

1) Enable http monitoring:
```
network_config:
  enabled: true
  enable_http_monitoring: true
```
2) Run the system-probe alone (without the rest of the datadog-agent). Verify you see `http monitoring has been enabled` in the logs
3) Make an http request:`curl http://www.httpbin.org/anything/qaisfun`
4) Wait a moment, then query the system-probe http debug endpoint: `Invoke-WebRequest 'http://localhost:3333/network_tracer/debug/http_monitoring' | Select-Object -Expand Content`
5) Verify that you see your query in the list

### Checklist
<!-- Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields. -->

- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] The `need-change/operator` and `need-change/helm` labels has been applied if applicable.
- [x] The appropriate `team/..` label has been applied, if known.
- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated if applicable.

Note: Adding GitHub labels is only possible for contributors with write access.

ISauve committed:
Add http monitoring for windows

brycekahle commented on pkg/network/driver/read_buffers.go:
```diff
@@ -0,0 +1,96 @@
+func PrepareCompletionBuffers(handle windows.Handle, count int) (iocp windows.Handle, buffers []*ReadBuffer, err error) {
+	iocp, err = windows.CreateIoCompletionPort(handle, windows.Handle(0), 0, 0)
+	if err != nil {
+		return windows.Handle(0), nil, errors.Wrap(err, "error creating IO completion port")
```
Can we use `fmt.Errorf` in this PR instead of `errors.Wrap`?

brycekahle commented on pkg/network/http/model_windows.go:
```diff
@@ -0,0 +1,119 @@
+	b := tx.RequestFragment
+
+	var start, end int
+	for start = 0; start < len(b) && b[start] != ' '; start++ {
```
See the gnuTLS PR for an example of how to solve this, but we need to handle a null byte before the end of the buffer. 

Also, this logic looks very similar to the Linux side. Is there a difference, or could we combine them?

ISauve replied:
The logic is basically identical to the Linux side, but the types are different (`[25]int8` on windows vs `[25]byte`), so on Linux we can use `copy` and on windows we have to copy & cast each byte ourselves.

brycekahle replied:
I realize this would require a driver change, but could we use an unsigned char type in the driver so both Linux and Windows could use `byte`? I think the `BYTE` type is available: https://docs.microsoft.com/en-us/windows/win32/winprog/windows-data-types

brycekahle commented on pkg/network/http/driver_interface.go:
```diff
@@ -0,0 +1,238 @@
+		nil,
+		uint32(0), nil, nil)
+	if err != nil {
+		log.Warnf("Failed to set max number of flows in driver http filter to %v %v", maxFlows, err)
```
Rather than have this always log, it feels like the caller should log the error, if they want. It seems the one place that calls this function doesn't handle the error either.

brycekahle commented on pkg/network/http/driver_interface.go:
```diff
@@ -0,0 +1,238 @@
+
+			err = driver.StartNextRead(di.driverHTTPHandle.Handle, buf)
+			if err != nil && err != windows.ERROR_IO_PENDING {
+				log.Infof("Error starting next http transaction read: %v")
```
Is `INFO` the appropriate log level? I think either going to the warn/error side, or the debug side makes sense here.

brycekahle commented on pkg/network/http/driver_interface.go:
```diff
@@ -0,0 +1,238 @@
+				return
+			}
+			if err != nil {
+				log.Infof("Error reading http transaction buffer: %v", err)
```
Check log level (see comment below)

brycekahle commented on pkg/network/http/driver_interface.go:
```diff
@@ -0,0 +1,238 @@
+		driver.FlushPendingHttpTxnsIOCTL,
+		nil, uint32(0), nil, uint32(0), nil, nil)
+	if err != nil {
+		log.Warnf("Failed to flush pending http transactions: %v", err)
```
I'd return the error here and log where it is called.

brycekahle commented on pkg/network/http/driver_interface.go:
```diff
@@ -0,0 +1,238 @@
+		for {
+			buf, bytesRead, err := driver.GetReadBufferWhenReady(di.iocp)
+			if iocpIsClosedError(err) {
+				return
```
Does this loop exiting cause the error to be surfaced elsewhere, or would we silently stop monitoring HTTP?

ISauve replied:
I'm not returning/surfacing this error because it's the intended exit condition for this loop - the IOCP won't close until we close the file handle which we've associated with it, so this should only occur when we call `httpDriverInterface.close`. I'll add a debug log line here so that the process isn't silent though

brycekahle commented on pkg/network/http/driver_interface.go:
```diff
@@ -0,0 +1,238 @@
+	// ERROR_OPERATION_ABORTED or ERROR_ABANDONED_WAIT_0 indicates that the iocp handle was closed
+	// during a call to GetQueuedCompletionStatus.
+	// ERROR_INVALID_HANDLE indicates that the handle was closed prior to the call being made.
+	return err == syscall.Errno(windows.ERROR_OPERATION_ABORTED) ||
```
Can we use `errors.Is(err, windows.ERROR_OPERATION_ABORTED)` for these checks in the PR? Also, is the cast necessary. It appears `windows.ERROR_OPERATION_ABORTED` is already a `syscall.Errno`?

brycekahle commented on pkg/network/http/monitor_windows.go:
```diff
@@ -0,0 +1,119 @@
+
+// Start consuming HTTP events
+func (m *Monitor) Start() {
+	if m == nil {
```
Why are we checking the pointer value in each function? It we are using a `nil` `Monitor` somewhere, that feels like a bug and should be fixed.

ISauve replied:
The httpMonitor is `nil` if http monitoring isn't enabled: https://github.com/DataDog/datadog-agent/blob/6db0da8cfbd833ff0a89c876d40a8722d9fa41a2/pkg/network/tracer/tracer_windows.go#L79-L86
FWIW, we do this on the linux side as well.

brycekahle replied:
I don't love that pattern, because it relies upon someone remembering to add that to every function. I'd rather take the interface+no-op implementation approach. If you add a function to the interface, you cannot forget to add it to the no-op version because compilation will fail.

ISauve replied:
I like that approach - I'll switch to that on both linux & windows

brycekahle commented on pkg/network/http/driver_interface.go:
```diff
@@ -0,0 +1,238 @@
+}
+
+func deepCopyTransactionData(dest, src *driver.HttpTransactionType) {
+	dest.Tup.Saddr = src.Tup.Saddr
+	dest.Tup.Daddr = src.Tup.Daddr
+	dest.Tup.Sport = src.Tup.Sport
+	dest.Tup.Dport = src.Tup.Dport
+	dest.Tup.Protocol = src.Tup.Protocol
+	dest.Tup.Family = src.Tup.Family
+	dest.Tup.Pid = src.Tup.Pid
```
`*dest = *src` would do a copy for you.

brycekahle commented on pkg/network/http/driver_interface.go:
```diff
@@ -0,0 +1,238 @@
+}
+
+func deepCopyTransactionData(dest, src *driver.HttpTransactionType) {
+	dest.Tup.Saddr = src.Tup.Saddr
+	dest.Tup.Daddr = src.Tup.Daddr
+	dest.Tup.Sport = src.Tup.Sport
+	dest.Tup.Dport = src.Tup.Dport
+	dest.Tup.Protocol = src.Tup.Protocol
+	dest.Tup.Family = src.Tup.Family
+	dest.Tup.Pid = src.Tup.Pid
+
+	dest.RequestStarted = src.RequestStarted
+	dest.ResponseLastSeen = src.ResponseLastSeen
+	dest.RequestMethod = src.RequestMethod
+	dest.ResponseStatusCode = src.ResponseStatusCode
+	dest.RequestFragment = src.RequestFragment
```
`*dest = *src` would do a copy for you.

ISauve replied:
oops 🤦  I'll blame this on too much switching between Go & C 😆

brycekahle commented on pkg/network/http/driver_interface.go:
```diff
@@ -0,0 +1,238 @@
+	go func() {
+		defer di.eventLoopWG.Done()
+
+		transactionSize := uint32(C.size_of_http_transaction_type())
```
Does `unsafe.Sizeof(C.HTTP_TRANSACTION_TYPE)` or `unsafe.Sizeof(HttpTransactionType)` work instead of this custom function?

brycekahle replied:
Or `C.sizeof_struct_HttpTransactionType`

ISauve replied:
Unfortunately not. This was the only solution I was able to find which compiles

brycekahle replied:
I had a typo, it should be `C.sizeof_struct__HttpTransactionType` (note the double underscore).

djmitche reviewed (approved):
release.json change looks good :)

djmitche reviewed (approved):
WINDOWS_DDNPM_DRIVER changes look good

derekwbrown committed:
Modifications for combining the HTTP processing path (in the driver)
with the regular flow stats processing

derekwbrown committed:
Agent (system probe) changes corresponding to driver updates for HTTP. (#12683)

Remove all of the overlapped processing for HTTP transactions.
Configure at probe startup max number of outstanding transactions,
number of transactions the driver should buffer before notifying the
probe, and the HTTP buffer length.

Have a single read loop (implemented with Ioctl()) for reading
transactions from the driver.

ISauve commented:
Closing because these changes have been pulled into #13103
